### DISCUSSION TITLE
Set Linux boot options (u-boot)?

### DISCUSSION THREAD

**Discussion by blhiggins** [1000001101000/Debian_on_Buffalo#158] [Q&A] [answered]
Hello!

I have a headless LS220D with Stretch installed (via Debian on Buffalo) that I'm now trying to upgrade to Buster, and then to Bullseye. Buster's release notes  mentions the network interface names changing, so I'm preparing for that. But using udevadm, it seems like the "eth0" name is coming from the hardware. Perhaps updating the devicetree would be enough to cause the names to change. Or if not that, I fear that once systemd 252 or later arrives (in an eventual upgrade to Bookworm) the names might change then (see systemd 252 release notes about "will now consult the Devicetree"). Either way is a problem if it means I can't SSH in after reboot.

So I was planning to add the "net.ifnames=0" to the Linux boot options. My understanding is this should force using the old interface name "eth0". That way, I can test at each upgrade stage whether udevadm suggests a "modern" interface name that I could migrate to (and then remove the "ifnames" boot option).

Now I realize the device uses u-boot (rather than the grub that I'm used to), and the boot options don't appear to be present in /boot. Can anyone me how the Linux boot options are changed? I'm a bit out of my depth at this point.

It's also possible my assessment of some aspect of this situation isn't right, so feel free to correct me if that's the case.

Best,
Brendon

**Comment by 1000001101000** [comment, marked as answer]
Good to hear from you!

I know that they still come through as eth0/1 up through Bullseye. I tried installing kernel 6.0 on a TS3400D the other day and found sata wasn't detected. This means I've got some kernel/device-tree work to do before I get very far into testing Bookworm, though I could probably generate an installer sooner than later which gets far enough to see what the interface name are even if it can't do a working install yet. I typically don't do too much with the new distro until the kernel version gets frozen to avoid spending too much time chasing bugs early on.

Even if the names change fresh installs via the installer should be fine since the interfaces are all autodetected etc. As you've pointed out, someone using my debootstrap script or doing a dist-upgrade could run into issues the naming changes for this setup. 

Changing the kernel command line is tricky on these devices, I generally recommend against it. You can adjust the mtd device that contains the uboot parameters to be RW and use the relevant uboot tools to change the cmdline but if you manage to mess something up you could end up needing a serial console to recover. I believe I've linked to some info from someone on the wiki who figured out a serial console on that model but it involves some intermediate soldering skill. It would be simple to append/replace it via DTB except that the relevant kernel option (CONFIG_ARM_ATAG_DTB_COMPAT_CMDLINE_EXTEND) is not enabled. 

There are a number of ways you could deal with this issue if/when the interface name changes, I imagine I'll need to document one of them when the time comes for others who do a dist-upgrade. 

The easiest might be to just add entries to the interfaces file for the new names once we know what they are. 
You could also create a script that determines what they are and deals with them via an initrd hook or even systemd service that runs before the network config. Another option is to switch to systemd-networkd which makes it really easy to do things like configure all e* interfaces to use DHCP. 

On a related note, I should probably point out I've made a lot of changes to the installer and related files over the years and have done a terrible job with backward-compatibility. I generally recommend folks re-run the installer rather than dist-upgrade as a result. Most don't apply for the ls220d but the script I might need to be updated to work properly. 

I hope I at least partially answered your question.

**Reply by blhiggins** [reply]
All of that insight is helpful, thank you! I think the major takeaways are that it's not straightforward to change the boot options, and I don't actually need to worry about this right now because the interface names haven't changed yet.

Should the interface names change in future, another possible option might be to (ab)use a systemd.link file. I was hesitant to go this route because I wasn't sure how stable the MAC would be, so I would have to match against everything. That should be fine for a single interface, like in my case, but doesn't seem like a terribly robust approach in general. It could work as a temporary stop-gap during an upgrade, though.

Thanks very much for your prompt help, and your work on this project!
